sparkler-plugins/fetcher-htmlunit/src/main/java/edu/usc/irds/sparkler/plugin/HtmlUnitFetcher.java

```diff
@@ -0,0 +1,147 @@
+        driver.setJavaScriptTimeout(DEFAULT_JS_TIMEOUT);
+        WebClientOptions options = driver.getOptions();
+        options.setCssEnabled(false);
+        options.setAppletEnabled(false);
+        options.setDownloadImages(false);
+        options.setJavaScriptEnabled(true);
+        options.setTimeout(DEFAULT_TIMEOUT);
+        options.setUseInsecureSSL(true);
+        options.setPopupBlockerEnabled(true);
+        options.setDoNotTrackEnabled(true);
+        options.setGeolocationEnabled(false);
+        options.setHistorySizeLimit(2);
+        options.setPrintContentOnFailingStatusCode(false);
+        options.setThrowExceptionOnScriptError(false);
+        options.setThrowExceptionOnFailingStatusCode(false);
+        if (this.httpHeaders != null && !this.httpHeaders.isEmpty()) {
+            LOG.info("Found {} headers", this.httpHeaders.size());
+            this.httpHeaders.forEach((name, val) -> driver.addRequestHeader(name, val));
+        } else {
+            LOG.info("No user headers found");
+        }
+    }
+
+    @Override
+    public FetchedData fetch(Resource resource) throws Exception {
+        LOG.info("HtmlUnit FETCHER {}", resource.getUrl());
+        FetchedData fetchedData;
+        try {
+            String userAgent = getUserAgent();
+            if (StringUtils.isNotBlank(userAgent)) {
+                driver.removeRequestHeader(USER_AGENT);
+                driver.addRequestHeader(USER_AGENT, userAgent);
+            }
+            Page page = driver.getPage(resource.getUrl());
+
+            WebResponse response = page.getWebResponse();
+            boolean truncated = false;
+            try (InputStream stream = response.getContentAsStream()) {
+                try (BoundedInputStream boundedStream = new BoundedInputStream(stream, CONTENT_LIMIT)) {
+                    try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+                        IOUtils.copy(boundedStream, out);
+                        fetchedData = new FetchedData(out.toByteArray(), response.getContentType(), response.getStatusCode());
+                        long contentLength = page.getWebResponse().getContentLength();
+                        if (contentLength > 0 && contentLength < Integer.MAX_VALUE) {
+                            fetchedData.setContentLength((int) contentLength);
+                            truncated = (contentLength > fetchedData.getContentLength());
+                        }
+                    }
+                }
+            }
+            resource.setStatus(ResourceStatus.FETCHED.toString());
+
+            List<NameValuePair> respHeaders = page.getWebResponse().getResponseHeaders();
+            Map<String, List<String>> headers = new HashMap<>();
+            fetchedData.setHeaders(headers);
+            if (respHeaders != null && !respHeaders.isEmpty()){
+                respHeaders.forEach(item -> {
+                    if (!headers.containsKey(item.getName())) {
```

Comment:
`respHeaders` is `List<NameValuePair>`. It is like List of tuples.
We are converting it to `Map<String, List<String>>`.

AFAIK, *some* http headers can be duplicative. so we are treating headervalues as list. For example, some stupid servers returns multiple set-cookie headers http://stackoverflow.com/q/3806132/1506477